src/Val.h

```diff
@@ -582,6 +582,8 @@ class StringVal final : public Val
 
 	StringValPtr Replace(RE_Matcher* re, const String& repl, bool do_all);
 
+	ValPtr FromJSON(const TypePtr& t) const;
```

Comment:
> Maybe a std::variant<ValPtr, std::string> where the string would be an error message?

I tried to implement this, but there is a small question. The original design was to continue parsing when encountering parsing errors, which means multiple errors may occur. If this design encounters an error, will it immediately return the error message and stop parse?

Reply:
> If this design encounters an error, will it immediately return the error message and stop parse?

Yes, I think that's fine. If there's an error, the user most likely cannot trust the result (at least can not do so in script-land).

Suggest to move the ValFromJSON method into the `zeek::detail` namespace so if it turns out we want to do really fancy stuff (think Django's per-field errrors responses) we can change without deprecation.

Reply:
Get it.

May we use std::exception in the internal function? like this:
```cpp
static ValPtr BuildVal(const rapidjson::Value& j, const TypePtr& t)
	{
        // ......
	throw std::runtime_error(util::fmt("type '%s' mismatch", type_name(t->Tag())));
	}

std::variant<zeek::ValPtr, std::string> ValFromJSON(std::string_view json_str, const TypePtr& t)
	{
        // ......
	try
		{
		return BuildVal(doc, t);
		}
	catch ( const std::runtime_error& e )
		{
		return e.what();
		}
	}
```
I think it could be more elegant, as I find using functional exception handling while recursively calling BuildVal in BuildVal may become too cumbersome.

Reply:
> I think it could be more elegant, as I find using functional exception handling while recursively calling BuildVal in BuildVal may become too cumbersome.

There are barely any exceptions in the code base, here's a small snippet (though the referencing counting is better with IntrusivePtr today).

https://github.com/zeek/zeek/wiki/Coding-Style-and-Conventions#exceptions

Could see either std::variant also for BuildVal or an actual `std::excepted` style helper, but yes, results in verbose "unwinding", but hopefully isn't all that bad.